[DM-28696] Representing missing values in the EFD

afausti opened the pull request:
- Discuss alternatives for representing missing values
- Describe the data flow from the moment a CSC publishes a record until the moment it is read by the user
- Show how InfluxDB handles missing values
- InfluxDB sink connector ability to skip NaN values
- InfluxDB schemaless property can be used to handle missing values
- A dropped NaN becomes NaN again in the Pandas df returned by the EFD client.

afausti committed:
Add a docker-compose setup to run InfluxDB

- Run the InfluxDB service to exercise the InfluxDB API in the technote.
- Ignore Influxdb data dir on git

afausti committed:
Start docker-compose services before build

afausti committed:
Install pandoc in CI

afausti committed:
Update requirements

Use documenteer >= 0.6.0 with modern version of sphinx and jupyter-sphinx

afausti committed:
Add the jupyter_sphinx extension

- With the jupyter_sphinx extension use the jupyter-execute directive to execute code cells and publish the cell outputs in your thechnote.
- Exclude files from sphinx build

afausti committed:
Initial draft

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+Missing values can occur in different ways:
+
+- **No value is available**. A situation that is often encountered during integration and testing is a DDS topic with a field defined in the topic schema, which is not implemented in the CSC yet. That happened when testing the Main Telescope Camera Cable Wrap subsystem recently. Another situation, which is expected to happen during operations, is when a sensor stops working.
+
```
I think it's worth making the case where a sensor goes off line another bullet in this list.  It's distinct from the case where a field in the schema is not implemented in that it represents a change in the functionality of the system.

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+- **A sensor produces invalid values**. For example, the Environment subsystem has a sensor that measures snow depth. The sensor may produce invalid values if there is no snow. In this case the CSC might represent those values as missing.
+
+In general, we need a representation for missing values when one or more field values in a topic are missing.
+There is still a timestamp, and eventually, other fields with valid values in the same topic to record in the EFD.
```
other fields --> there may be other fields

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+In this technote, we discuss alternatives for representing missing values.
+We review the data flow from the moment a CSC publishes data until the user analyses it.
+In particular, we show how the different systems involved can handle missing values.
+Finally, we propose a solution for representing missing values that works globally and ensures we can correctly read the missing values using the EFD client, the supported method for accessing EFD data.
```
Maybe provide a link to https://efd-client.lsst.io/?

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+
+`SAL Kafka`_ and `InfluxDB Sink`_ are of particular importance because they are responsible for converting the records between the data formats used by the different systems.
+SAL Kafka converts the DDS records into Avro records and forwards them to Kafka.
+It is also responsible for converting the IDL schema to Avro schema and register them with the Kafka schema registry.
```
and register --> and registering

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+    +--------------------------------------------------------------+
+
+For the EFD, telemetry is where a representation for missing values is more important.
+From Table 1, the majority of the telemetry fields (83%) has ``float`` or ``double`` IDL types.
```
has --> have

Also, you might mention that these are bolded in the table because they are the dominant types.  When I looked at the table, I thought they were bolded because they were totals.  Perhaps it would be handy to also have a line for totals?

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+
+In Table 1, arrays are counted only once. When the topic field is an array, we show the type of the array items.
+
+Table 2 shows the number of arrays in the ``ts_xml`` schema per array size in Telemetry topics. The largest arrays are from ``MTCamera``, ``MTM1M3`` subsystems and usually have ``float`` or ``double`` IDL types. However, there are fairly large arrays in the ``MTM1M3TS`` and ``MTM2`` subsystems with ``boolean``, ``unsigned short`` or ``long`` IDL types.
```
largest arrays are from --> largest arrays are from the

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+.. _DDS type system: https://community.rti.com/static/documentation/connext-dds/6.0.0/doc/manuals/connext_dds/getting_started_extras/RTI_ConnextDDS_CoreLibraries_GettingStarted_ExtensibleTypesAddendum.pdf
 
-  and
+By default, DDS initializes or "unset"  optional fields to ``NULL``.
```
"unset" --> "unsets"

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+If this field is omited by the CSC that publishes this topic, DDS would automatically fill in the default value ``NaN`` for the application that subscribes to it.
 
-.. note::
+By default, the default value of an enumeration corresponds to the first value in the enumeration.
```
Having default twice here makes it a bit hard to read.  Maybe say: "The default behavior for enumerations is to assign the first value in the enumeration as the default value."

Or I think you could just omit the first i.e. "By default, the default" --> "The default"

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+  };
+
+Currently, in ``ts_xml``, it is not possible to set optional fields or explicitly set default values for fields.
+An application that subscribes to a DDS topic will get what the CSCs publish or the default values presented in Table 3.
```
publish --> publishes

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+    BLUE
+  };
+
+Currently, in ``ts_xml``, it is not possible to set optional fields or explicitly set default values for fields.
```
I'm afraid I don't understand the implication of this sentence.  It's not possible for what to set optional fields?

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+    "name": "foo",
+    "fields" : [
+      {"name": "bar", "type": "float"},
+      {"name": "baz", "type": "float", "default": NaN} // field baz default to NaN
```
Why `NaN` here and `null` above.

Also, I think I'm missing something the first block is an optional field and I understand why you'd want a default, but the second one is not optional, right?  In that case doesn't that mean it has to show up.  If so, what is the point of a default?

Edit: I guess you are saying we could translate directly between an IDL schema and an Avro schema

afausti replied:
Agreed, I'm not convinced that the use of optinal fields is equivalent in DDS and Avro. I don't think optional fields add much here. I'll restrict the discussion to using default values, and show that it is possible to translate an IDL schema to an Avro schema in this case.

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+
+From this and previous sections, SAL Kafka could, in principal, convert an IDL schema with optional and default fields into an Avro schema.
+
+A field with the value ``NaN`` in a DDS record passed along to the Avro record, and that has been tested with SAL Kafka.
```
I'm missing the sign on this.  You are saying a field with the value `NaN` *can* be passed along to the Avro record?  I think it just needs a few more words to make that explicit

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+
+  <measurement>[,<tag_key>=<tag_value>[,<tag_key>=<tag_value>]] <field_key>=<field_value> [,<field_key>=<field_value>] [<timestamp>]
+
+In the EFD, a **mesasurement** corresponds to a DDS topic, and **tags** are metadata associated to **fields**. In InfluxDB, we specify a **timestamp** in the Unix epoch. If a timestamp is not specified, InfluxDB uses the server's local time with nanoseconds precision.
```
metadata associated to --> metadata associated with
nanoseconds precision --> nanosecond precision

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+The Python keyword ``None`` is used here as the default value for an optional parameter (the InfluxDB field in this case), as expected.
+
+The InfluxDB Sink connector is the right place for dropping missing values. In particular, the ``influxdb-java`` library used by the connector, `skips fields with NaN values`_ when writing to InfluxDB.
+This feature was implemented upstream a month after we first notice this problem back in September 2019, during the AuxTel integration activities at the Summit.
```
notice this problem --> noticed this problem

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+.. _writing NaN values to InfluxDB: https://jira.lsstcorp.org/browse/DM-21300
+
+
+In summary, fields are optional, and we should drop missing values when writing to InfluxDB.
```
Maybe you are going to cover this later on, but this still doesn't address the issue that we don't know if a field is missing because it was never filled in or because it was intentionally set to NaN.

afausti replied:
It is just designed as missing in the EFD. I don't think we can distinguish these two situations.  I'll clarify this further in the introduction when we describe the situations when missing values can occur.

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+
+For sparse arrays with ``NaN`` values the connector will extract them and drop the corresponding fields with ``NaN`` values before writing to InfluxDB.
+
+For an Avro record with an array ``[1.0, NaN, 2.0]``, for example,  the connector will first extract the array items into the field set ``foo1=1.0,foo2=NaN,foo3=2.0`` and then write ``foo1=1.0,foo3=3.0`` to InfluxDB.
```
I didn't realize this could happen.  I don't think the efd client will handle this gracefully now if some elements of an array are missing.

afausti replied:
Notice you still can query `foo2` and the missing value should get converted back to `NaN` by the EFD client. I'll try to show that in the following section.

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+
+
+EFD client
+==========
```
This section needs some fleshing out.

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+
+- Show that missing values in the InfluxDB are converted back to NaN when the EFD client returns a Pandas dataframe.
+
+- Point to the documentation on how to handle missing data in Pandas, and say why NaNs are more convenient than convenient
```
Are you arguing NaNs are convenient, or inconvenient?

afausti replied:
Added the missing section and clarified this point.

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
+Using a sentinel value like ``NaN`` to represent missing values works for ``float`` and ``double`` IDL types as a partial solution.
+A field with the value ``NaN`` in a DDS record is converted to a field with value ``NaN`` in the Avro record,  and it is dropped by the InfluxDB Sink connector.
+In particular, we showed that when querying InfluxDB in Python, it fills the missing values with ``None``.
+When using the EFD client, missing values are converted back to ``NaN`` which is a convenient way to represent missing values with Pandas ``float64`` dtype.
```
with Pandas --> with the Pandas

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
 
-.. Add content here.
-.. Do not include the document title (it's automatically added from metadata.yaml).
+This solution is currently in place, and it solves our problem for about 80% of the fields in telemetry topics. Our recommendation for DDS topics with sparse arrays of ``float`` or ``double`` IDL types is to fill in the missing values also with ``NaN`` values.
```
values also with --> values with

SimonKrughoff commented on index.rst:
```diff
@@ -1,61 +1,461 @@
 
-.. .. bibliography:: local.bib lsstbib/books.bib lsstbib/lsst.bib lsstbib/lsst-dm.bib lsstbib/refs.bib lsstbib/refs_ads.bib
-..    :style: lsst_aa
+As a final recommendation, we should consider using the ``@default`` annotation in the IDL schema to set the default values for ``float`` and ``double`` fields to ``NaN`` instead of ``0``.
```
Oh, I think this should be much more front and center.  I didn't appreciate that 0 was the default for float and double until I got to here!  Maybe we should put a bullet list of recommendations right at the beginning.

afausti replied:
Agreed. I'll add a new section for recomendation to make this more explicit.

SimonKrughoff reviewed (changes requested):
My main comment is that we have defaults that work reasonably for all datatypes, but for float and double the values always look like they are missing, but they could have been intentionally set to NaN.
For the other datatypes, they are always recorded so look like they were intentionally set to the default value, and we can't tell when they are actually missing.

That distinction is address obliquely, but I think we should make that clear.  I understand we can't solve that problem generically right now, although I do wonder if we could do something with a tag saying which type a particular field is, but that would require schema changes and we would still have to record some sort of sentinel value in InfluxDB which is always dangerous.

I think the above should probably be addressed before we merge it.

Most of the rest of my comments are trivial

afausti commented:
@SimonKrughoff that's great feedback! I agree with you reading again after some time it's hard to grasp the main conclusions. 

Thanks for the detailed comments, I plan on working on it tomorrow again. I think that adding an abstract to highlight the main points will help.  I'll restrict the discussion on representing missing values and having a default value for each type, and remove the text around defining optional fields which is confusing. I'll add the missing session on the EFD client as well. Initially my intention was to stop at InfluxDB but then in the last minute decided to include the EFD client to present the end-to-end picture of the problem. I feel this technote will be much improved after your comments.

afausti committed:
Add the aioinflux and pandas dependencies

- They are used to demonstrate that a missing value in InfluxDB is returned back as NaN in a Pandas dataframe

afausti commented:
@SimonKrughoff I have addressed most of your review comments.
